sdks/python/apache_beam/examples/wordcount.py

```diff
@@ -85,31 +98,66 @@ def run(argv=None):
   pipeline_options.view_as(SetupOptions).save_main_session = True
   p = beam.Pipeline(options=pipeline_options)
 
+  reveal_type(p)
+
+  _read = ReadFromText(known_args.input)
+  reveal_type(_read)
+
+  read = 'read' >> _read
+  reveal_type(read)
+
   # Read the text file[pattern] into a PCollection.
-  lines = p | 'read' >> ReadFromText(known_args.input)
+  lines = p | read
+  reveal_type(lines)  # PCollection[unicode*]
+
+  def make_ones(x):
+    # type: (T) -> Tuple[T, int]
```

Comment:
Can this not be inferred? Similarly below.

Reply:
Can you explain what you mean?   

Let me try to answer what I think you're asking about.

First, even for a simple function like `make_ones`, mypy does not guess at the arguments and return type.  

Next, as a user annotating pipelines I have a few of choices  (all of them are valid):

1. annotate functions, and let mypy infer the types of the `PTransforms` they are passed to
2. annotate `PTransforms`
3. annotate both, as a way to guard against unintentional incompatibilities

Option 2 would look like this:

```python
  def make_ones(x):
    return (x, 1)

  makemap = beam.Map(make_ones)  # type: beam.Map[T, Tuple[T, int]]
```

To be clear, in option 2, mypy is not using the type of `beam.Map` to infer the type of the `make_ones`, it's simply assuming it's correct, since in our thought experiment it's unannotated.  In other words, the following would not generate an error because `make_ones` is untyped:

```python
  def make_ones(x):
    return (x + 1, 1)  # x can only be a numeric type!

  makemap = beam.Map(make_ones)  # type: beam.Map[str, Tuple[str, int]]
```

This module is very messy right now, so I apologize for any confusion that it's generated.

Reply:
That's really unfortunate... I'd have expected more of mypy :(. In that case our own inference (which does handle this) is all the more important.